Lib/gettext.py

```diff
@@ -655,3 +737,144 @@ def npgettext(context, msgid1, msgid2, n):
 # gettext.
 
 Catalog = translation
+
+
+# utils for t-string handling in gettext translation + pygettext extraction
+# TBD where they should go, and whether this should be a public API or internal,
+# especially the part about generating names from interpolations which is IMHO
+# beneficial to have in stdlib so any implementation can re-use it without
+# risking diverging behavior for the same expression between implementations
+
+class _NameTooComplexError(ValueError):
+    """
+    Raised when an expression is too complex to derive a format string name
+    from it, or the resulting name would not be valid in a format string.
+    """
+
+
+class _ExtractNamesVisitor(ast.NodeVisitor):
+    def __init__(self):
+        self._name_parts = []
+
+    @property
+    def name(self) -> str:
+        name = '__'.join(self._name_parts)
```

Comment:
I used `__` as a separator between parts since initially I thought that it might be nice to make it clearer that the placeholder for `{user.name}` isn't just something named `user_name`.

However, maybe using a single underscore would be fine here:

- `t'{user.name} {user_name}'` would simply fail due to the check that a name doesn't map to different expressions
- I can't come up with a good example where you would use `foo.bar` and `foo_bar` in the same string